File: go/solutions/0001_two_sum/two_sum.go
package two_sum

// 注意本题返回的是数组的下标，因此不能排序后用双指针处理，这会导致索引丢失

// 解法一：暴力解法
// Time: O(n^2), Space: O(1)
func twoSumBruteForce(nums []int, target int) []int {
	for i := range nums {
		for j := i + 1; j < len(nums); j++ {
			if nums[i]+nums[j] == target {
				return []int{i, j}
			}
		}
	}

	return []int{}
}

// 解法二：哈希表+一次遍历
// Time: O(N), Space: O(N)
func twoSumHashMap(nums []int, target int) []int {
	seen := make(map[int]int, len(nums))
	for i, num := range nums {
		if j, ok := seen[target-num]; ok {
			return []int{i, j}
		}

		seen[num] = i
	}

	return []int{}
}

File: python/solutions/0001_two_sum/two_sum.py
class Solution:
    def twoSumBruteForce(self, nums: list[int], target: int) -> list[int]:
        n = len(nums)
        for i in range(n):
            for j in range(i + 1, n):
                if nums[i] + nums[j] == target:
                    return [i, j]

        return []

    # twoSumHashMap
    def twoSumHashMap(self, nums: list[int], target: int) -> list[int]:
        seen: dict[int, int] = {}
        for i, num in enumerate(nums):
            if target - num in seen:
                return [i, seen[target - num]]
            seen[num] = i
        return []
